### ISSUE TITLE
IPython notebook hangs

### ISSUE THREAD

**Issue by dhermes** [blink1073/oct2py#61]
Running the provided [sample](http://nbviewer.ipython.org/github/blink1073/oct2py/blob/master/example/octavemagic_extension.ipynb) causes my entire machine (not just the example) to hang when using the subplot examples.

Machine info:

```

$ ipython --version
2.3.0
$ pip show oct2py

---
Name: oct2py
Version: 2.4.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: 
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.1 LTS"
$ octave --version
GNU Octave, version 3.8.1
...
```

**Comment by blink1073** [comment]
Hi @dhermes, would you care to try this with the latest version (3.5) and see if it works for you?

**Comment by josephcslater** [comment]
Sorry @dhermes didn't reply. I'm in a similar scenario to him and still seeing a hang. I saw your instructions above, and got it unstuck using -g. However, now it matches the [failure that I get in my linux installation](https://gist.github.com/josephcslater/86718ae919248ba355fe9587327e67ad). 

Thoughts? (Please be a bit patient- I'm likely to only be able to test a solution this weekend).

**Comment by blink1073** [comment]
@josephcslater, I'll look at this after we get https://github.com/Calysto/octave_kernel/issues/50 fixed.

**Comment by josephcslater** [comment]
Thanks. Let me know if I can test anything.

**Comment by blink1073** [comment]
@josephcslater, your issue should be fixed by upgrading oct2py to v3.5.10.

**Comment by josephcslater** [comment]
Unfortunately, it's still hanging. 

```
%load_ext oct2py.ipython
%%octave
a = 1:10
disp(a)
```
Just hangs. 

Likewise
```
x = %octave [1 2; 3 4];
x
```
hangs.

**Comment by blink1073** [comment]
I just tried it again on Ubuntu 14.04, and it is working fine, can you verify the following?  

```python
In [1]: import oct2py; oct2py.__version__
Out[1]: '3.5.10'
```

**Comment by josephcslater** [comment]
Confirmed. 

Works fine on my Ubuntu installation as well. Failing on my Mac.

**Comment by blink1073** [comment]
Ah, I had to follow the "Using Aquaterm with Gnuplot" from  http://wiki.octave.org/Octave_for_MacOS_X to get it to work on my Mac.

**Comment by josephcslater** [comment]
I'm not even trying to plot. Should this matter?

**Comment by blink1073** [comment]
Yes because oct2py doesn't know if you are trying to plot until after you run the command, so we treat every command as a possible plotting command.

**Comment by josephcslater** [comment]
Hmmm, so this may not work for me. What I have working is octave 4.0.3 installed using macports (not brew). It plots via xterm, not aquaterm . Not sure where to go from here.

**Comment by josephcslater** [comment]
I don't think this is my problem. I just ran gnu plot, changed the terminal to aqua (`set term aqua`) and plotting followed fine. So, gnu plot will recognize aqua. It may be that I need to set the default to aqua for oct2py interacting with it to work?

**Comment by blink1073** [comment]
That shouldn't be required, AFAIK.   Are you able to plot from `octave-cli` without changing any settings?

**Comment by josephcslater** [comment]
Yes.

**Comment by josephcslater** [comment]
At least this isn't (so far?) the face plant of my issue #50 (I owe you a beer for that level of stupid on my part!).

**Comment by blink1073** [comment]
I'm not entirely sure where to go from here.  I am not keen to install Macports and spending another round of debugging the inner loop at the moment.

**Comment by josephcslater** [comment]
I get it. Have an idea how I could trace it?

**Comment by josephcslater** [comment]
It's not a strong need of mine right now, but more of a "this would be nice to have", so I appreciate how much you've already put into it.

**Comment by blink1073** [comment]
You could try putting print statements in the inner `evaluate` function, that is essentially what I ended up doing this morning to debug the Homebrew plotting.  I'll be afk for the rest of the day.

**Comment by josephcslater** [comment]
Thank you, and enjoy the time away.

**Comment by blink1073** [comment]
Okay, I think the problem is trying to do control flow logic (for loops in this case) through the pipe, which is too brittle.   v3.6.0 adds a separate `_make_figs.m` function that encapsulates the logic.     Do you mind testing that release?

**Comment by josephcslater** [comment]
Works! Thanks! 

That you would ask if I mind... I owe it to you (and other freeware authors (which I am as well) to do so. Thank you for coming back to this. 

FWIW: The failure happened even when no figures were involved. Further, figures are indeed working. 

In light of this: I thought I should test your demo ipynb, it unfortunately failed on the cell

```
%%octave -s 500,500

# butterworth filter, order 2, cutoff pi/2 radians
b = [0.292893218813452  0.585786437626905  0.292893218813452];
a = [1  0  0.171572875253810];
freqz(b, a, 32);
```

the fail results being:

```
---------------------------------------------------------------------------
ExpatError                                Traceback (most recent call last)
<ipython-input-10-bca0f2444242> in <module>()
----> 1 get_ipython().run_cell_magic('octave', '-s 500,500', '\n# butterworth filter, order 2, cutoff pi/2 radians\nb = [0.292893218813452  0.585786437626905  0.292893218813452];\na = [1  0  0.171572875253810];\nfreqz(b, a, 32);')

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/IPython/core/interactiveshell.py in run_cell_magic(self, magic_name, line, cell)
   2113             magic_arg_s = self.var_expand(line, stack_depth)
   2114             with self.builtin_trap:
-> 2115                 result = fn(magic_arg_s, cell)
   2116             return result
   2117 

<decorator-gen-124> in octave(self, line, cell, local_ns)

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    186     # but it's overkill for just that one bit of state.
    187     def magic_deco(arg):
--> 188         call = lambda f, *a, **k: f(*a, **k)
    189 
    190         if callable(arg):

/Users/jslater/Library/Python/3.5/lib/python/site-packages/oct2py/ipython/octavemagic.py in octave(self, line, cell, local_ns)
    337             if plot_format == 'svg':
    338                 image = self._fix_gnuplot_svg_size(image, size=(plot_width,
--> 339                                                                 plot_height))
    340             display_data.append((key, {plot_mime_type: image}))
    341 

/Users/jslater/Library/Python/3.5/lib/python/site-packages/oct2py/ipython/octavemagic.py in _fix_gnuplot_svg_size(self, image, size)
    109 
    110         """
--> 111         (svg,) = minidom.parseString(image).getElementsByTagName('svg')
    112         viewbox = svg.getAttribute('viewBox').split(' ')
    113 

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/xml/dom/minidom.py in parseString(string, parser)
   1966     if parser is None:
   1967         from xml.dom import expatbuilder
-> 1968         return expatbuilder.parseString(string)
   1969     else:
   1970         from xml.dom import pulldom

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/xml/dom/expatbuilder.py in parseString(string, namespaces)
    923     else:
    924         builder = ExpatBuilder()
--> 925     return builder.parseString(string)
    926 
    927 

/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/xml/dom/expatbuilder.py in parseString(self, string)
    221         parser = self.getParser()
    222         try:
--> 223             parser.Parse(string, True)
    224             self._setup_subset(string)
    225         except ParseEscape:

ExpatError: not well-formed (invalid token): line 189, column 92
```

**Comment by josephcslater** [comment]
Now hanging is:

```
%%octave -s 600,200 -f png
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```

**Comment by josephcslater** [comment]
The final cell (using the octave plotting GUI) works fine. 

Of course, order matters. It can't be executed with the others hanging.

**Comment by josephcslater** [comment]
Also hanging is:

```
%%octave -f svg
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```

However, it works fine with the `-g` option.

**Comment by blink1073** [comment]
Thanks again for the reports @josephcslater, I'll try to sneak in some time to work on this over the holiday.

**Comment by josephcslater** [comment]
It's hard. I know. I can't even get nose tests to work on one of my packages on travis-ci. I know (believe) the package works, but I'd like to eventually be able to rely on nose tests to determine release. Programing gremlins are the worst time sinks.

**Comment by blink1073** [comment]
I tracked down the issue to gnuplot creating an invalid SVG file in this case:

![image](https://cloud.githubusercontent.com/assets/2096628/20569213/ea2d5666-b164-11e6-983c-db70eb4c5ad0.png)

**Comment by josephcslater** [comment]
egad.  I suppose the only solution is a [bug report](https://sourceforge.net/p/gnuplot/bugs/) to gnuplot.

**Comment by blink1073** [comment]
I have a work-around coming, it ends up loading as such (notice the error character in the caption):

![image](https://cloud.githubusercontent.com/assets/2096628/20569798/f5d78084-b166-11e6-8daa-61e33f873132.png)

**Comment by blink1073** [comment]
7581f39 has the fix.  The `sombrero` example is showing no output at all for me on OSX, I'll look at that next.

**Comment by blink1073** [comment]
I released 3.6.1 that should address the problems you saw.

**Comment by josephcslater** [comment]
I dread saying this, but indeed that plot works, but 
```%%octave -s 600,200 -f png
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
``` 
hangs now, along with the subsequent `%%octave -f svg cell.

**Comment by blink1073** [comment]
The second cell will never execute if the first hangs.  Are you able to execute both if you change the first to:

```
%%octave -s 600,200 -f svg
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```

**Comment by josephcslater** [comment]
"The second cell will never execute if the first hangs. Are you able to execute both if you change the first to:"

Of course. I restarted the entire notebook to test the second cell, skipping the one that had hung.  

Tested, and indeed, the answer is unfortunately "no."

**Comment by blink1073** [comment]
Let's try and rule this out from an Octave session:

```
subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
print(gcf, 'test.png')
print(gcf, 'test.svg')
```

**Comment by josephcslater** [comment]
There are three ways for me to launch: from terminal, `octave` or `octave-cli`, of double clicking in the GUI. 

The first two work fine as long as I'm launching from a directory I can write to. If not, it hangs. I got to that thought only after running from the GUI and realizing it defaults to `/`, which is non-user writable. 

That seems like it.  The issue might be that octave is launching with pwd set to `/`. The trick may very well be to set that to `/tmp` or pretty much anything else (if only to test). After that, of course, in the software design process you want the file saved to some appropriate place, which is likely the directory that `jupyter notebook` was launched from. I have no idea how easy/hard this is. It's either trivial or a pain, but that depends on how your code works (which is a bit of voodoo to me!).

**Comment by blink1073** [comment]
We can change directory to your home directory at startup with this change (note it adds `cd` to the list of commands on that line).  Does making this change allow your notebook to run?

```diff
diff --git a/oct2py/core.py b/oct2py/core.py
index e4bf3f5..3846679 100644
--- a/oct2py/core.py
+++ b/oct2py/core.py
@@ -860,7 +860,7 @@ class _Session(object):
     def _handle_first_run(self):
         here = os.path.realpath(os.path.dirname(__file__))
         here = here.replace(os.path.sep, '/')
-        self.write('disp(available_graphics_toolkits());more off;addpath(genpath("%s"));disp(char(3))\n' % here)
+        self.write('disp(available_graphics_toolkits());more off;cd;addpath(genpath("%s"));disp(char(3))\n' % (here, here))
         resp = self.expect(chr(3))
         if not os.name == 'nt':
             try:
```

**Comment by josephcslater** [comment]
I presume there is a simple way to apply this, I'm going to manually edit the file `core.py` for the sake of expediency. Excuse my barbarism.

**Comment by blink1073** [comment]
Hack away ;).  Applying a git diff is not high on the list of intuitive tasks to do ;).

**Comment by josephcslater** [comment]
I've applied such before... but don't trust myself. 

Nevertheless, `:(`. This failed. I also tried to explicitly put in a path that would be clearly writable. To my desktop, with and without training `/`.

**Comment by josephcslater** [comment]
```
%%octave  -f svg
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```
works, 
but 

```
%%octave -s 600,200 -f svg
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
``` 
fails. 

Clearly the `-s` option is the issue.

**Comment by josephcslater** [comment]
`%%octave  -f svg -s 600,200` (swapping order of arguments) did not help.

**Comment by blink1073** [comment]
That is *weird*, because the height and width are getting set either way even if you don't specify them.

**Comment by josephcslater** [comment]
The fact that what I said makes no sense did not elude me.

**Comment by josephcslater** [comment]
So, as a piece of information that I can't understand, 

```
%%octave -s 500,500
# Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```
doesn't work, but

```
%%octave -s 500,500

# butterworth filter, order 2, cutoff pi/2 radians
b = [0.292893218813452  0.585786437626905  0.292893218813452];
a = [1  0  0.171572875253810];
freqz(b, a, 32);
``` 
does.

**Comment by josephcslater** [comment]
Ready to flip out? I've made progress (dramatic pause).

**Comment by josephcslater** [comment]
The comment line:
```
# Note: On Windows, this will not show the plots unless Ghostscript is installed.
```
needed to be removed.

**Comment by josephcslater** [comment]
The `png` example is still failing... I'm working on editing the cell itself to see of the comment line is an issue.

**Comment by josephcslater** [comment]
```
%%octave -s 600,200 -f png

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```
works, but 

```
%%octave -s 600,200 -f png

% Note: On Windows, this will not show the plots unless Ghostscript is installed.

subplot(121);
[x, y] = meshgrid(0:0.1:3);
r = sin(x - 0.5).^2 + cos(y - 0.5).^2;
surf(x, y, r);

subplot(122);
sombrero()
```
doesn't. 

lmao.

**Comment by josephcslater** [comment]
So, it seems that a comment line before you get to code causes a problem, regardless of using `#` or `%`. 

In the middle: unchecked. I think at the end doesn't matter (last example) because the work is already done.

**Comment by josephcslater** [comment]
Of course, that's apparently not always true, as the buterworth example has been working fine all along. 

Removing the comment in that example does not fix the odd character issue, btw. I thought I'd try, though I didn't expect it to.

**Comment by josephcslater** [comment]
So... 

for the butterworth example,
```%%octave  -f svg -s 200,600```

does to change the size at all. 

Neither does

```%%octave -s 200,600 ```
versus 
```%%octave -s 500,500 ```

It seems the first line parsing has some odd issue that it's ignoring something, including getting messed up when there is a subsequent comment line.

**Comment by josephcslater** [comment]
Some extreme numbers in sizing show that it does something, but I think something in Jupyter is rescaling, and Octave might be doing something similar. It's hard to find a value to it's use right now watching what I'm seeing. The result is coming through clean by default.

**Comment by josephcslater** [comment]
(tyfys)

**Comment by blink1073** [comment]
We have two problems: our ability to differentiate between a `%magic` and a `% comment`, and yes, I did notice that our ability to set the size is wonky.   I am going to take a break for this for a while, at least we made some progress ;).

**Comment by josephcslater** [comment]
It is also `# comments` that have problems. Easy enough to work around (avoid) for the user, but programming may be impossible as it's possible jupyter chews up anything after a % so you are stuck with it feeding `oct2py` comment garbage that you can't interpret. 

I'm betting your wife is giving you a hard time. I'm in work, and this is more fun that what I need to be doing. You taking a break may save my marriage.

**Comment by blink1073** [comment]
Right, Octave supports either kind of comment character:  https://www.gnu.org/software/octave/doc/v4.0.3/Single-Line-Comments.html

This could be related to how we're evaluating the input, I have been meaning to switch to the method used by pymatbridge.  Their method of setting the image size is also better.  The image size handling I can add soon, the other my family would not appreciate right now ;).

**Comment by josephcslater** [comment]
I might need a couch tonight the way my day is going.

**Comment by josephcslater** [comment]
I don't need anything quickly on this, so don't do it for me.

**Comment by blink1073** [comment]
`v3.6.2` should fix the size wonkiness.  Thanks for all the detailed follow-ups :).

**Comment by josephcslater** [comment]
Everything seems to work, sizing seems better (I have't thoroughly tested), but comments still break it. (even in the 3.6.3 release).

**Comment by josephcslater** [comment]
FWIW: Something is broken with this package display on `pypi`. 
![untitled](https://cloud.githubusercontent.com/assets/12815335/20641184/530f9dbc-b3c0-11e6-9009-31b50af79a9f.png)

I'm likely to try to use oct2py as a template for my packages [array_to_latex](https://pypi.python.org/pypi/array_to_latex/0.36) and [vibration_toolbox](https://pypi.python.org/pypi/vibration_toolbox/0.5b9), so helping out helps me as well. 

Right now I feel like they are held together with duct tape. I definitely feel like a monkey banging on them!

**Comment by blink1073** [comment]
Thanks for the heads up, I removed the badge.  Yup, writing libraries is hard.  This was my first one, and I'm still learning after 5 years ;).

**Comment by josephcslater** [comment]
My career path has taken me far from this, but coding is my outlet and helps me feel like I'm still relevant in modern computing. Can I ask- for the life of me I can't see how the [oct2py web pages](http://blink1073.github.io/oct2py/) end up being served on github. I didn't find the repository, I see the make files and all, but in my head I would have to build at my terminal to an html out directory that I would also have defined as a repository on github. It's exactly how I built [vibrationtoolbox.github.io](vibrationtoolbox.github.io). I'd like to to better going forward, but boy it seems mystifying to me. Your `.travis.yml` file gave me some thoughts on some of my issues, though. Sorry for detouring from the thread, but I have no idea where else to ask.

**Comment by blink1073** [comment]
By default, the github.io site points to the index.html in the root directory of the gh-pages branch of the repo.  Recently they added the ability to point the page to other locations, in the settings for the repo.

**Comment by josephcslater** [comment]
My mind just got blown. 

So, the pages are in the sources directory, but the make file makes the pages and puts them in a separate branch somehow? I think in this. 

```
gh-pages:
	git checkout master
	git pull origin master
	cp oct2py/tests/*.m example
	git commit -a -m "Keep examples in sync"; true
	git push origin; true
	make docs
	ghp-import -n -p -m $(GHP_MSG) docs/_build/html

```

This will take a while to deconstruct.

**Comment by blink1073** [comment]
The last two lines are all you should need.  It builds the docs and uses the ghp-import python package to publish automagically.  See the docs for the package for what the options mean.
